drv/gimlet-seq-server/src/main.rs

```diff
@@ -103,19 +104,56 @@ enum Trace {
     SpdBankAbsent(u8),
     SpdAbsent(u8, u8, u8),
     SpdDimmsFound(usize),
-
+    I2cFault {
+        retries_remaining: u8,
+        code: i2c::ResponseCode,
+    },
+    StartFailed(SeqError),
     None,
 }
 
 ringbuf!(Trace, 128, Trace::None);
 
 #[export_name = "main"]
 fn main() -> ! {
-    let spi = drv_spi_api::Spi::from(SPI.get_task_id());
     let sys = sys_api::Sys::from(SYS.get_task_id());
     let jefe = Jefe::from(JEFE.get_task_id());
+    let spi = drv_spi_api::Spi::from(SPI.get_task_id());
     let hf = hf_api::HostFlash::from(HF.get_task_id());
+    match init(&sys, &jefe, spi, hf) {
+        // Set up everything nicely, time to start serving incoming messages.
+        Ok(mut server) => {
+            let mut buffer = [0; idl::INCOMING_SIZE];
+            loop {
+                idol_runtime::dispatch(&mut buffer, &mut server);
+            }
+        }
 
+        // Initializing the sequencer failed.
+        Err(_) => {
+            // Tell everyone that something's broken, as loudly as possible.
+            ringbuf_entry!(Trace::StartFailed(SeqError::I2cFault));
+            sys.gpio_set(FAULT_PIN_L);
+            jefe.set_state(PowerState::InitFailed as u32);
+
+            // All these moments will be lost in time, like tears in rain...
+            // Time to die.
+            loop {
+                hl::sleep_for(100);
```

Comment:
The IPC server probably assumes that initialization was successful, so it's unsafe to start it here... that's too bad because it'd be really great to be able to ask it questions.

We'd have to introduce some notion of "I'm broken" into the IPC server for that to work.

Mostly just thinking out loud here.

Reply:
Yeah, I was also thinking about possibly having it start serving, and just respond to everything with an "I'm broken" message. Would be happy to change that now, if you think that's a good idea, or in a later PR?